File: tests/org.eclipse.ice.io.test/src/org/eclipse/ice/io/hdf/test/HdfReaderFactoryTester.java
/*******************************************************************************
 * Copyright (c) 2013, 2014 UT-Battelle, LLC.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *   Initial API and implementation and/or initial documentation - Jay Jay Billings,
 *   Jordan H. Deyton, Dasha Gorin, Alexander J. McCaskey, Taylor Patterson,
 *   Claire Saunders, Matthew Wang, Anna Wojtowicz
 *******************************************************************************/
package org.eclipse.ice.io.hdf.test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;
import org.eclipse.ice.io.hdf.HdfFileFactory;
import org.eclipse.ice.io.hdf.HdfReaderFactory;

import java.io.File;
import java.net.URI;
import java.util.ArrayList;

import ncsa.hdf.object.Attribute;
import ncsa.hdf.object.Dataset;
import ncsa.hdf.object.Datatype;
import ncsa.hdf.object.Group;
import ncsa.hdf.object.HObject;
import ncsa.hdf.object.h5.H5Datatype;
import ncsa.hdf.object.h5.H5File;
import ncsa.hdf.object.h5.H5Group;
import ncsa.hdf.object.h5.H5ScalarDS;

import org.junit.BeforeClass;
import org.junit.Ignore;
import org.junit.Test;

/**
 * <!-- begin-UML-doc -->
 * <p>
 * This class tests the HdfReaderFactory class.
 * </p>
 * <!-- end-UML-doc -->
 * 
 * @author els
 * @generated 
 *            "UML to Java (com.ibm.xtools.transform.uml2.java5.internal.UML2JavaTransform)"
 */
public class HdfReaderFactoryTester {
	/**
	 * <!-- begin-UML-doc -->
	 * <p>
	 * This operation conducts any required initialization for the tests.
	 * </p>
	 * <!-- end-UML-doc -->
	 * 
	 * @generated 
	 *            "UML to Java (com.ibm.xtools.transform.uml2.java5.internal.UML2JavaTransform)"
	 */
	@BeforeClass
	public static void beforeClass() {
		// begin-user-code

		// end-user-code
	}

	/**
	 * <!-- begin-UML-doc -->
	 * <p>
	 * This operation checks the getChildH5Group, getChildH5Groups, and
	 * getDataset operations.
	 * </p>
	 * <!-- end-UML-doc -->
	 * 
	 * @generated 
	 *            "UML to Java (com.ibm.xtools.transform.uml2.java5.internal.UML2JavaTransform)"
	 */
	@Test
	public void checkGetters() {
		// begin-user-code

		// Local declarations
		String separator = System.getProperty("file.separator");
		String testFileName = "hdfReaderFactoryFile1.h5";
		File dataFile = new File(System.getProperty("user.dir") + separator
				+ "data" + separator + testFileName);
		URI uri = dataFile.toURI();
		H5File h5File;
		String subGroupName = "Bob";
		H5Group subH5Group = null;
		ArrayList<String> rows = new ArrayList<String>();
		int maxRowSize = 0;
		String rowName = "Rows";

		// Setup ArrayList of Rows
		rows.add("ABBA");
		rows.add("BABBA");
		rows.add("CABBA");
		rows.add("DabbbyDabbyDooo!");
		rows.add("E");

		// Setup the string array
		String[] rowArray = new String[rows.size()];

		// Figure out the biggest row size and add to array
		for (int i = 0; i < rows.size(); i++) {
			maxRowSize = Math.max(maxRowSize, rows.get(i).length());
			// Add to array
			rowArray[i] = rows.get(i);
		}

		// Create a good H5File
		h5File = HdfFileFactory.createH5File(uri);
		assertNotNull(h5File);

		// Open file and setup H5 File to have a parent and child. Check get
		// Readers and get Datasets
		try {
			h5File.open();
			// Setup the H5 File to have a Parent and child
			// Get the parent group
			H5Group parentH5Group = (H5Group) ((javax.swing.tree.DefaultMutableTreeNode) h5File
					.getRootNode()).getUserObject();

			// Test to see if a getChild returns null - group does not exist
			assertNull(HdfReaderFactory.getChildH5Group(parentH5Group,
					subGroupName));

			// Null Checks
			assertNull(HdfReaderFactory.getChildH5Group(null, subGroupName));
			assertNull(HdfReaderFactory.getChildH5Group(parentH5Group, null));
			assertNull(HdfReaderFactory.getChildH5Group(null, null));

			// Add a group to the root group
			h5File.createGroup(subGroupName, parentH5Group);

			// Persist data
			h5File.close();
			h5File.open();

			// Get the parent
			parentH5Group = (H5Group) ((javax.swing.tree.DefaultMutableTreeNode) h5File
					.getRootNode()).getUserObject();

			// Check to see it exists with the group grabber
			subH5Group = HdfReaderFactory.getChildH5Group(parentH5Group,
					subGroupName);

			// Check the subH5Groups name to make sure that it does exist and
			// has the correct name
			assertNotNull(subH5Group);
			assertTrue(subGroupName.equals(subH5Group.getName()));

			// Check getDatasets

			// Check invalid usage of getting dataset that does not exist
			assertNull(HdfReaderFactory.getDataset(subH5Group, rowName));

			// Nullary checks for dataSet
			assertNull(HdfReaderFactory.getDataset(null, rowName));
			assertNull(HdfReaderFactory.getDataset(subH5Group, null));
			assertNull(HdfReaderFactory.getDataset(null, null));

			// Add a Dataset to the subH5Group

			// Create datatypes for the row and column labels
			H5Datatype rowH5Datatype = (H5Datatype) h5File.createDatatype(
					Datatype.CLASS_STRING, maxRowSize, Datatype.NATIVE,
					Datatype.NATIVE);
			long[] dims = { rows.size() };

			// Add array to dataset and add dataset ot the subgroup
			Dataset h5RowDataset = h5File.createScalarDS(rowName, subH5Group,
					rowH5Datatype, dims, null, null, 0, null);
			h5RowDataset.write(rowArray);

			// Close group and then reopen
			try {
				h5File.close();
				h5File.open();
			} catch (Exception e1) {
				e1.printStackTrace();
				dataFile.delete();
				fail();
			}

			// Get the group again
			parentH5Group = (H5Group) ((javax.swing.tree.DefaultMutableTreeNode) h5File
					.getRootNode()).getUserObject();

			subH5Group = (H5Group) parentH5Group.getMemberList().get(0);

			// Now, check to see if you can get it!
			Dataset data = HdfReaderFactory.getDataset(subH5Group, rowName);
			// Make sure it is not null
			assertNotNull(data);

			// Check information on dataset
			assertEquals(1, subH5Group.getMemberList().size());
			// Check the value of that dataset
			assertEquals(rowName, subH5Group.getMemberList().get(0).getName());
			// Check labels that they are not a group
			assertFalse(subH5Group.getMemberList().get(0) instanceof Group);

			// Check DataSet information

			// Check the rowData
			Object rowData = data.getData();

			// Iterate over the list and check the arrayLists
			for (int i = 0; i < rows.size(); i++) {
				// Check contents of rowData at X
				assertEquals(rows.get(i), ((String[]) rowData)[i]);
			}

			// Check the getter for index
			// Check nullaries
			assertNull(HdfReaderFactory.getChildH5Group(null, 0));
			assertNull(HdfReaderFactory.getChildH5Group(parentH5Group, -1));
			assertNull(HdfReaderFactory.getChildH5Group(null, -1));
			// Check invalid index - out of bounds
			assertNull(HdfReaderFactory.getChildH5Group(parentH5Group, 1));

			// Check valid usage
			subH5Group = HdfReaderFactory.getChildH5Group(parentH5Group, 0);

			// Check the subH5Groups name to make sure that it does exist and
			// has the correct name
			assertNotNull(subH5Group);
			assertEquals(subGroupName, subH5Group.getName());

			// Try to grab the group as a dataset - should throw a class cast
			// exception
			assertNull(HdfReaderFactory.getDataset(parentH5Group, subGroupName));

			// Try to grab the dataset as a group - should throw a class cast
			// exception
			assertNull(HdfReaderFactory.getChildH5Group(subH5Group, rowName));
			assertNull(HdfReaderFactory.getChildH5Group(subH5Group, 0));

		} catch (Exception e) {
			// Delete the dataFile
			dataFile.delete();
			e.printStackTrace();
			fail();
		}

		// Delete file
		dataFile.delete();

		// end-user-code
	}

	/**
	 * <!-- begin-UML-doc -->
	 * <p>
	 * This operation checks the readDoubleAttribute, readStringAttribute, and
	 * readIntegerAttribute operations.
	 * </p>
	 * <!-- end-UML-doc -->
	 * 
	 * @generated 
	 *            "UML to Java (com.ibm.xtools.transform.uml2.java5.internal.UML2JavaTransform)"
	 */
	@Test
	public void checkReaders() {
		// begin-user-code

		// Local declarations
		String separator = System.getProperty("file.separator");
		String testFileName = "hdfReaderFactoryFile2.h5";
		File dataFile = new File(System.getProperty("user.dir") + separator
				+ "data" + separator + testFileName);
		URI uri = dataFile.toURI();
		H5File h5File = null;
		String integerName = "Integer1";
		String doubleName = "Double1";
		String stringName = "String1";
		int intValue = 3;
		double doubleValue = 5.67656565343;
		String stringValue = "Bobby Lee Jones";
		String childGroupName = "new Group";

		// Create a good H5File
		h5File = HdfFileFactory.createH5File(uri);
		assertNotNull(h5File);

		// Open file and setup H5 File to have a parent and child. Check get
		// Readers and get Datasets
		try {
			h5File.open();
			// Setup the H5 File to have a Parent and child
			// Get the parent group
			H5Group parentH5Group = (H5Group) ((javax.swing.tree.DefaultMutableTreeNode) h5File
					.getRootNode()).getUserObject();

			// Check nullary reader operations

			// Check nullary and invalid values for Integer read
			assertNull(HdfReaderFactory.readIntegerAttribute(parentH5Group,
					integerName));
			assertNull(HdfReaderFactory.readIntegerAttribute(null, integerName));
			assertNull(HdfReaderFactory.readIntegerAttribute(parentH5Group,
					null));
			assertNull(HdfReaderFactory.readIntegerAttribute(null, null));

			// Check nullary and invalid values for Double read
			assertNull(HdfReaderFactory.readDoubleAttribute(parentH5Group,
					doubleName));
			assertNull(HdfReaderFactory.readDoubleAttribute(null, doubleName));
			assertNull(HdfReaderFactory
					.readDoubleAttribute(parentH5Group, null));
			assertNull(HdfReaderFactory.readDoubleAttribute(null, null));

			// Check nullary and invalid values for String read
			assertNull(HdfReaderFactory.readStringAttribute(parentH5Group,
					stringName));
			assertNull(HdfReaderFactory.readStringAttribute(null, stringName));
			assertNull(HdfReaderFactory
					.readStringAttribute(parentH5Group, null));
			assertNull(HdfReaderFactory.readStringAttribute(null, null));

			// Add a Integer, a Double, and a String to the parentGroup

			// Setup dataTypes for Integer, Double, and a String

			// Integer Datatype
			// Create the integer datatype
			H5Datatype intDatatype = (H5Datatype) h5File.createDatatype(
					Datatype.CLASS_INTEGER, Datatype.NATIVE, Datatype.NATIVE,
					Datatype.NATIVE);

			// Double (float) Datatype
			// Create the float datatype
			H5Datatype doubleDatatype = (H5Datatype) h5File.createDatatype(
					Datatype.CLASS_FLOAT, 8, Datatype.NATIVE, Datatype.NATIVE);

			// String Datatype
			// Create a custom String data type for the value
			H5Datatype stringDatatype = (H5Datatype) h5File.createDatatype(
					Datatype.CLASS_STRING, stringValue.length(),
					Datatype.NATIVE, Datatype.NATIVE);

			// Add a Integer
			long[] intDims = { 1 };
			int[] intValues = { intValue };
			// Create an attribute object
			Attribute intAttribute = new Attribute(integerName, intDatatype,
					intDims, intValues);
			// Write the attribute
			parentH5Group.writeMetadata(intAttribute);

			// Add a Double
			long[] doubleDims = { 1 };
			double[] doubleValues = { doubleValue };
			// Create an attribute object
			Attribute doubleAttribute = new Attribute(doubleName,
					doubleDatatype, doubleDims, doubleValues);
			// Write the attribute
			parentH5Group.writeMetadata(doubleAttribute);

			// Add a String
			long[] stringDims = { 1 };
			String[] stringValues = new String[1];
			stringValues[0] = stringValue;
			// Create a byte array from values using the stringToByte method
			// See
			// http://mail.hdfgroup.org/pipermail/hdf-forum_hdfgroup.org/2011-March/004509.html
			byte[] bvalue = Dataset.stringToByte(stringValues,
					stringValue.length());
			// Create an attribute object
			Attribute stringAttribute = new Attribute(stringName,
					stringDatatype, new long[] { 1 });
			// Set the value of the attribute to bvalue
			stringAttribute.setValue(bvalue);
			// Write the attribute to the group's metadata
			parentH5Group.writeMetadata(stringAttribute);

			h5File.close();

			h5File.open();
			parentH5Group = (H5Group) ((javax.swing.tree.DefaultMutableTreeNode) h5File
					.getRootNode()).getUserObject();

			// Now, time to read those values!
			assertEquals(
					intValue,
					HdfReaderFactory.readIntegerAttribute(parentH5Group,
							integerName).intValue());
			assertEquals(
					doubleValue,
					HdfReaderFactory.readDoubleAttribute(parentH5Group,
							doubleName).doubleValue(), 0.0);
			assertEquals(stringValue, HdfReaderFactory.readStringAttribute(
					parentH5Group, stringName));

			// Checking for child groups.
			assertNotNull(HdfReaderFactory.getChildH5Groups(parentH5Group));
			assertEquals(0, HdfReaderFactory.getChildH5Groups(parentH5Group)
					.size());

			// Check for child objects.
			assertNotNull(HdfReaderFactory.getChildH5Members(parentH5Group));
			assertEquals(0, HdfReaderFactory.getChildH5Members(parentH5Group)
					.size());

			// Lets add something to the group
			H5Group h5Group = (H5Group) h5File.createGroup(childGroupName,
					parentH5Group);
			// Add an attribute to the group
			// Add a Integer
			h5Group.writeMetadata(intAttribute);
			assertNotNull(h5Group);

			// Add a dataset (100x50 integer array) to the parent group.
			String datasetName = "2D integer";
			Datatype dtype = new H5Datatype(Datatype.CLASS_INTEGER, 4,
					Datatype.ORDER_LE, Datatype.SIGN_NONE);
			long[] dims = { 100, 50 };
			long[] maxdims = dims;
			long[] chunks = null;
			int gzip = 0;
			Object data = null;
			h5File.createScalarDS(datasetName, parentH5Group, dtype, dims,
					maxdims, chunks, gzip, data);

			// Close to persist the data and reopen, get the new parenth5Group
			// object
			h5File.close();

			h5File.open();
			parentH5Group = (H5Group) ((javax.swing.tree.DefaultMutableTreeNode) h5File
					.getRootNode()).getUserObject();

			// Check for children now. There should only be one child group.
			assertNotNull(HdfReaderFactory.getChildH5Groups(parentH5Group));
			assertEquals(1, HdfReaderFactory.getChildH5Groups(parentH5Group)
					.size());
			assertEquals(childGroupName,
					HdfReaderFactory.getChildH5Groups(parentH5Group).get(0)
							.getName());

			// Check for child objects. There should be two (a group and a
			// dataset).
			assertNotNull(HdfReaderFactory.getChildH5Members(parentH5Group));
			assertEquals(2, HdfReaderFactory.getChildH5Members(parentH5Group)
					.size());
			assertEquals(childGroupName,
					HdfReaderFactory.getChildH5Members(parentH5Group).get(1)
							.getName());
			assertEquals(datasetName,
					HdfReaderFactory.getChildH5Members(parentH5Group).get(0)
							.getName());

			// Delete the file
			dataFile.delete();

		} catch (Exception e) {
			// Fail out of the test, not supposed to happen
			e.printStackTrace();
			dataFile.delete(); // Delete the dataFile
			fail();
		}

		// end-user-code
	}
}

File: eclipse-initial/tests/native/org.eclipse.ice.io.native.test/hdf_test/HdfReaderFactoryTester.cpp
/*******************************************************************************
 * Copyright (c) 2012, 2014 UT-Battelle, LLC.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *   Initial API and implementation and/or initial documentation - Jay Jay Billings,
 *   Jordan H. Deyton, Dasha Gorin, Alexander J. McCaskey, Taylor Patterson,
 *   Claire Saunders, Matthew Wang, Anna Wojtowicz
 *******************************************************************************/
#define BOOST_TEST_DYN_LINK
#define BOOST_TEST_MODULE Regression
#include <boost/test/included/unit_test.hpp>
#include <HdfReaderFactory.h>
#include <HdfWriterFactory.h>
#include <HdfFileFactory.h>
#include <memory>
#include <string>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/types.h>

using namespace ICE_IO;

BOOST_AUTO_TEST_SUITE(HdfReaderFactoryTester_testSuite)



BOOST_AUTO_TEST_CASE(beforeClass) {
    //Create folder testPath in local directory - relative!
    std::string testFolder = "ICEIOTestsDir/";
    mkdir(testFolder.c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);

    //Delete file
    std::string testFile = testFolder + "test.h5";
    remove(testFile.c_str());
}
BOOST_AUTO_TEST_CASE(checkGetters) {

    // begin-user-code

    //Local Declarations
    std::string testFileName = "test1.h5";
    std::string testFolder = "ICEIOTestsDir/";
    std::string dataFile = testFolder + testFileName +"";
    std::shared_ptr<H5::H5File> h5File;
    std::shared_ptr<H5::Group> badGroup;
    std::shared_ptr<H5::H5File> badFile;
    double doubleNumber = 1.2131231231231;
    int intNumber = 33;
    std::string words = "Man, C++ is so much FUN!";

    //Used for exception hitting
    bool exceptionHit0 = false;
    bool exceptionHit1 = false;
    bool exceptionHit2 = false;
    bool exceptionHit3 = false;
    bool exceptionHit4 = false;
    bool exceptionHit5 = false;
    bool exceptionHit6 = false;

    std::string subGroupName = "Bob";
    std::vector<std::string> rows;
    int maxRowSize = 0;
    std::string rowName = "Rows";

    //Setup ArrayList of Rows
    rows.push_back("ABBA");
    rows.push_back("BABBA");
    rows.push_back("CABBA");
    rows.push_back("DabbbyDabbyDooo!");
    rows.push_back("E");

    //Setup the string array
    char *rowArray[rows.size()];

    //Figure out the biggest row size and add to array
    for (int i = 0; i < rows.size(); i++) {

        //Get the max length
        if(rows.at(i).length() > maxRowSize) {
            maxRowSize = rows.at(i).length();
        }
        //Add to array
        rowArray[i] = new char[rows.at(i).size() +1];
        std::strcpy(rowArray[i], rows.at(i).c_str());
    }

    //Create a good H5File
    h5File = HdfFileFactory::createH5File(dataFile);

    //Open file and setup H5 File to have a parent and child.  Check get Groups and get Datasets
    try {

        //Nullary Checks
        try {
            HdfReaderFactory::getChildH5Group(badGroup,subGroupName);
        } catch(...) {
            exceptionHit0 = true;
        }
        BOOST_REQUIRE_EQUAL(exceptionHit0, true);

        //Nullary Checks
        try {
            HdfReaderFactory::getChildH5Group(badGroup,0);
        } catch(...) {
            exceptionHit1 = true;
        }
        BOOST_REQUIRE_EQUAL(exceptionHit1, true);

        //Add a group
        std::unique_ptr<H5::Group> parentH5Group(new H5::Group(h5File.get()->createGroup("/Reactor")));

        //Add subGroups
        parentH5Group.get()->createGroup("Assembly1");
        parentH5Group.get()->createGroup("Assembly2");
        parentH5Group.get()->createGroup("Assembly3");

        //Create a dataset
        hid_t filetype = H5Tcopy(H5T_C_S1);
        H5Tset_size(filetype, H5T_VARIABLE);
        hid_t memtype = H5Tcopy(H5T_C_S1);
        H5Tset_size(memtype, H5T_VARIABLE);
        hsize_t dims[1] = { rows.size() };

        hid_t space = H5Screate_simple(1, dims, NULL);

        hid_t dset = H5Dcreate(parentH5Group.get()->getId(),"DS1", filetype, space, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
        H5Dwrite (dset, memtype, H5S_ALL, H5S_ALL, H5P_DEFAULT, rowArray);

        //Persist data
        parentH5Group.get()->close();
        h5File.get()->close();

        //Memory management!
        H5Tclose (filetype);
        H5Tclose (memtype);
        H5Sclose (space);
        H5Dclose (dset);

        h5File = HdfFileFactory::openH5File(dataFile);

        //Get the root Group
        std::shared_ptr<H5::Group> rootGroup(new H5::Group(h5File.get()->openGroup("/")));

        //Use the root group to get the parent group
        std::shared_ptr<H5::Group> parentGroup = HdfReaderFactory::getChildH5Group(rootGroup, "Reactor");


        //Nullary Checks - out of bounds
        try {
            HdfReaderFactory::getChildH5Group(parentGroup,10);
        } catch(...) {
            exceptionHit2 = true;
        }
        BOOST_REQUIRE_EQUAL(exceptionHit2, true);

        //Nullary Checks - negative index
        try {
            HdfReaderFactory::getChildH5Group(parentGroup, -1);
        } catch(...) {
            exceptionHit3 = true;
        }
        BOOST_REQUIRE_EQUAL(exceptionHit3, true);

        //Child does not exist
        try {
            HdfReaderFactory::getChildH5Group(parentGroup,subGroupName);
        } catch(...) {
            exceptionHit4 = true;
        }
        BOOST_REQUIRE_EQUAL(exceptionHit4, true);

        //Get the groups on the main group
        std::vector< std::shared_ptr<H5::Group> > childGroups = HdfReaderFactory::getChildH5Groups(parentGroup);

        //Make sure there are 3
        //There is a limit here because I am unable to get name types to compare at this time.
        //For now, this is a limitation on the CPP libraries for H5 and will have to be unsafely
        //assumed to be Groups and that they are valid groups.

        //FIXME - Need more test info for this piece.
        BOOST_REQUIRE_EQUAL(3, childGroups.size());

        //Get by index and by name and get by index to verify that they exist
        std::shared_ptr< H5::Group> assembly1a= HdfReaderFactory::getChildH5Group(parentGroup, 0);
        std::shared_ptr< H5::Group> assembly2a = HdfReaderFactory::getChildH5Group(parentGroup, 1);
        std::shared_ptr< H5::Group> assembly3a = HdfReaderFactory::getChildH5Group(parentGroup, 2);

        //Verify by name
        std::shared_ptr< H5::Group> assembly1b = HdfReaderFactory::getChildH5Group(parentGroup, "Assembly1");
        std::shared_ptr< H5::Group> assembly2b = HdfReaderFactory::getChildH5Group(parentGroup, "Assembly2");
        std::shared_ptr< H5::Group> assembly3b = HdfReaderFactory::getChildH5Group(parentGroup, "Assembly3");

        //Check dataset
        std::shared_ptr<H5::DataSet> dataSet = HdfReaderFactory::getDataset(parentGroup, "DS1");

        //Dataset Check - bad dataset
        try {
            std::shared_ptr<H5::DataSet> dataSet1 = HdfReaderFactory::getDataset(parentGroup, "notadataset");
        } catch (...) {
            exceptionHit5 = true;
        }
        BOOST_REQUIRE_EQUAL(exceptionHit5, true);

        //Dataset Check - bad group
        try {
            std::shared_ptr<H5::DataSet> dataSet1 = HdfReaderFactory::getDataset(badGroup, "DS1");
        } catch (...) {
            exceptionHit6 = true;
        }
        BOOST_REQUIRE_EQUAL(exceptionHit6, true);

        //Open up dataset and read values
        filetype = H5Dget_type(dataSet.get()->getId());
        space = H5Dget_space(dataSet.get()->getId());
        int ndims = H5Sget_simple_extent_dims(space, dims, NULL);
        char *data[dims[0]];

        memtype = H5Tcopy(H5T_C_S1);
        H5Tset_size(memtype, H5T_VARIABLE);
        H5Dread(dataSet.get()->getId(), memtype, H5S_ALL, H5S_ALL, H5P_DEFAULT, data);

        //Output data
        int i = 0;
        for(i = 0; i < dims[0]; i++) {
            BOOST_REQUIRE_EQUAL(rows.at(i), data[i]);
        }

        //Make sure the sizes are equal
        BOOST_REQUIRE_EQUAL(rows.size(), i);

        //Free the arrays
        for(i = 0; i < dims[0]; i++) {
            delete data[i];
            delete rowArray[i];
        }

    } catch (...) {
        BOOST_FAIL("HdfReaderFactory test failure! Exiting.");
    }

    return;
    // end-user-code

}
BOOST_AUTO_TEST_CASE(checkReaders) {

    // begin-user-code

    //Local Declarations
    std::string testFileName = "test1.h5";
    std::string testFolder = "ICEIOTestsDir/";
    std::string dataFile = testFolder + testFileName +"";
    std::shared_ptr<H5::H5File> h5File;
    std::shared_ptr<H5::Group> badGroup;
    std::shared_ptr<H5::H5File> badFile;
    double doubleNumber = 1.2131231231231;
    int intNumber = 33;
    std::string words = "Man, C++ is so much FUN!";

    std::string intName = "integer value";
    std::string doubleName = "double value";
    std::string strName = "string value";

    //Used for exception hitting
    bool exceptionHit0 = false;
    bool exceptionHit1 = false;
    bool exceptionHit2 = false;
    bool exceptionHit3 = false;
    bool exceptionHit4 = false;
    bool exceptionHit5 = false;
    bool exceptionHit6 = false;

    //Create a good H5File
    h5File = HdfFileFactory::createH5File(dataFile);

    //Open file and setup H5 File to have a parent and child.  Check Readers
    try {

        //Exception handling for errors

        //Check readInteger - bad group
        try {
            HdfReaderFactory::readIntegerAttribute(badGroup, intName);
        } catch(...) {
            exceptionHit0 = true;
        }
        BOOST_REQUIRE_EQUAL(exceptionHit0, true);

        //Check readDouble - bad group
        try {
            HdfReaderFactory::readDoubleAttribute(badGroup, doubleName);
        } catch(...) {
            exceptionHit1 = true;
        }
        BOOST_REQUIRE_EQUAL(exceptionHit1, true);

        //Check readString - bad group
        try {
            HdfReaderFactory::readStringAttribute(badGroup, strName);
        } catch(...) {
            exceptionHit2 = true;
        }
        BOOST_REQUIRE_EQUAL(exceptionHit2, true);

        //create a root group
        std::shared_ptr <H5::Group> rootGroup (new H5::Group(h5File.get()->createGroup("/Reactor")));

        //Error check operations on Reactor Group for values that do not exist

        //Check readInteger - no value exists for name
        try {
            HdfReaderFactory::readIntegerAttribute(rootGroup, intName);
        } catch(...) {
            exceptionHit3 = true;
        }
        BOOST_REQUIRE_EQUAL(exceptionHit3, true);

        //Check readDouble - no value exists for name
        try {
            HdfReaderFactory::readDoubleAttribute(rootGroup, doubleName);
        } catch(...) {
            exceptionHit4 = true;
        }
        BOOST_REQUIRE_EQUAL(exceptionHit4, true);

        //Check readString - no value exists for name
        try {
            HdfReaderFactory::readStringAttribute(rootGroup, strName);
        } catch(...) {
            exceptionHit5 = true;
        }
        BOOST_REQUIRE_EQUAL(exceptionHit5, true);

        //Create an integer, double, and string attribute on rootGroup.  Then persist file.

        //Create the dataspace
        hid_t space_id = H5Screate(H5S_SIMPLE);
        hsize_t dims[1] = { 1 };
        H5Sset_extent_simple(space_id, 1, dims, dims);

        //Create the attribute and write it
        //Create the attribute
        hid_t attribute1 = H5Acreate(rootGroup.get()->getId(), intName.c_str(), HdfWriterFactory::createIntegerH5Datatype(h5File).getId(), space_id, H5P_DEFAULT, H5P_DEFAULT);
        H5Awrite(attribute1, HdfWriterFactory::createIntegerH5Datatype(h5File).getId(), &intNumber);

        hid_t attribute2 = H5Acreate(rootGroup.get()->getId(), doubleName.c_str(), HdfWriterFactory::createFloatH5Datatype(h5File).getId(), space_id, H5P_DEFAULT, H5P_DEFAULT);
        H5Awrite(attribute2, HdfWriterFactory::createFloatH5Datatype(h5File).getId(), &doubleNumber);

        //Create the dataspace and create the attribute
        H5::StrType dataType(0, words.length());

        hid_t attribute3 = H5Acreate(rootGroup.get()->getId(), strName.c_str(), dataType.getId(), space_id, H5P_DEFAULT, H5P_DEFAULT);
        H5Awrite(attribute3, dataType.getId(), words.c_str());

        //Close file

        H5Aclose(attribute1);
        H5Aclose(attribute2);
        H5Aclose(attribute3);
        rootGroup.get()->close();
        h5File.get()->close();

        //Reopen file
        h5File = HdfFileFactory::openH5File(dataFile);

        //Get rootH5Group
        std::shared_ptr<H5::Group> rootH5Group(new H5::Group(h5File.get()->openGroup("/Reactor")));

        //Check values
        BOOST_REQUIRE_EQUAL(intNumber, HdfReaderFactory::readIntegerAttribute(rootH5Group, intName));
        BOOST_REQUIRE_EQUAL(doubleNumber, HdfReaderFactory::readDoubleAttribute(rootH5Group, doubleName));
        BOOST_REQUIRE_EQUAL(words, HdfReaderFactory::readStringAttribute(rootH5Group, strName));


    } catch (...) {
        BOOST_FAIL("HdfReaderFactory test failure! Exiting.");
    }

    return;
    // end-user-code

}

//Removes file and folder after test completes
BOOST_AUTO_TEST_CASE(afterClass) {

    //relative!
    std::string testFolder = "ICEIOTestsDir/";
    std::string testFile = testFolder + "test.h5";

    //Delete file
    remove(testFile.c_str());

    //Delete folder
    remove(testFolder.c_str());
}

BOOST_AUTO_TEST_SUITE_END()
